File: bob/vm.py
#-------------------------------------------------------------------------------
# bob: vm.py
#
# Implementation of the Bob virtual machine.
#
# Eli Bendersky (eliben@gmail.com)
# This code is in the public domain
#-------------------------------------------------------------------------------
from __future__ import print_function
from .bytecode import (
        OP_CONST, OP_LOADVAR, OP_STOREVAR, OP_DEFVAR, OP_FUNCTION, OP_POP,
        OP_JUMP, OP_FJUMP, OP_RETURN, OP_CALL)
from .expr import expr_repr, Boolean
from .builtins import BuiltinProcedure, builtins_map
from .environment import Environment
from .utils import Stack


DEBUG = False


class Closure(object):
    """ A closure is a code object (procedure) with an associated environment
        in which the closure was defined.
    """
    def __init__(self, codeobject, env):
        self.codeobject = codeobject
        self.env = env


class ExecutionFrame(object):
    """ Encapsulates the execution "frame": state of the VM. 
    
        codeobject: 
            The current code object executed by the VM
        pc: 
            An index into the code object of the next instruction to execute
        env: 
            The environment in which the code is being executed
    """
    def __init__(self, codeobject, pc, env):
        self.codeobject = codeobject
        self.pc = pc
        self.env = env


class BobVM(object):
    """ Implementation of the Bob VM. Initialize and then .run() with a 
        CodeObject.
    """
    class VMError(Exception): pass
        
    def __init__(self, output_stream=None):
        self.valuestack = Stack()
        self.framestack = Stack()
        
        self.frame = ExecutionFrame(
                        codeobject=None,
                        pc=None,
                        env=self._create_global_env())
        
        if output_stream is None:
            import sys
            self.output_stream = sys.stdout
        else:
            self.output_stream = output_stream
    
    def run(self, codeobject):
        """ The main execution function of the VM. Accepts a CodeObject, and
            runs it until there are no more instructions to execute.
        """
        self.frame.codeobject = codeobject
        self.frame.pc = 0
        
        #
        # The big VM loop!
        #
        while True:
            # If there are no more instructions in this code object, it's either 
            # because we're done executing the program, or an error.
            #
            instr = self._get_next_instruction()
            
            if instr is None:
                if self._is_in_toplevel_code():
                    break
                else:
                    raise self.VMError('Code object ended prematurely: %s' % self.codeobject)
            
            if DEBUG: print(opcode2str(instr.opcode), instr.arg)
            
            if instr.opcode == OP_CONST:
                self.valuestack.push(self.frame.codeobject.constants[instr.arg])
            elif instr.opcode == OP_LOADVAR:
                value = self.frame.env.lookup_var(self.frame.codeobject.varnames[instr.arg])
                self.valuestack.push(value)
            elif instr.opcode == OP_STOREVAR:
                value = self.valuestack.pop()
                self.frame.env.set_var_value(self.frame.codeobject.varnames[instr.arg], value)
            elif instr.opcode == OP_DEFVAR:
                value = self.valuestack.pop()
                self.frame.env.define_var(self.frame.codeobject.varnames[instr.arg], value)
            elif instr.opcode == OP_POP:
                if len(self.valuestack) > 0:
                    self.valuestack.pop()
            elif instr.opcode == OP_JUMP:
                self.frame.pc = instr.arg
            elif instr.opcode == OP_FJUMP:
                predicate = self.valuestack.pop()
                if isinstance(predicate, Boolean) and predicate.value is False:
                    self.frame.pc = instr.arg
            elif instr.opcode == OP_FUNCTION:
                func_codeobject = self.frame.codeobject.constants[instr.arg]
                closure = Closure(func_codeobject, self.frame.env)
                self.valuestack.push(closure)
            elif instr.opcode == OP_CALL:
                # For CALL what we have on TOS the function and then the 
                # arguments to pass to it - the last argument is highest on
                # the stack.
                # The function is either a BuiltinProcedure or a Closure (for
                # user-defined procedures)
                #
                proc = self.valuestack.pop()
                argvalues = [self.valuestack.pop() for i in range(instr.arg)]
                argvalues.reverse()
                
                if isinstance(proc, BuiltinProcedure):
                    result = proc.apply(argvalues)
                    self.valuestack.push(result)
                elif isinstance(proc, Closure):
                    if len(proc.codeobject.args) != len(argvalues):
                        raise self.VMError('Calling procedure %s with %s args, expected %s' % (
                                                proc.codeobject.name, len(args), len(proc.codeobject.args)))
                    
                    # We're now going to execute a code object, so save the
                    # current execution frame on the frame stack.
                    #
                    self.framestack.push(self.frame)

                    # Extend the closure's environment with the bindings of
                    # argument names --> passed values. 
                    #
                    arg_bindings = {}
                    for i, argname in enumerate(proc.codeobject.args):
                        arg_bindings[argname] = argvalues[i]
                    extended_env = Environment(arg_bindings, proc.env)
                    
                    # Start executing the procedure
                    #
                    self.frame = ExecutionFrame(
                                    codeobject=proc.codeobject,
                                    pc=0,
                                    env=extended_env)
                else:
                    raise self.VMError('Invalid object on TOS for CALL: %s' % proc)
                
            elif instr.opcode == OP_RETURN:
                self.frame = self.framestack.pop()
            else:
                raise self.VMError('Unknown instruction opcode: %s' % instr.opcode)
    
    def _get_next_instruction(self):
        """ Get the next instruction from the current code object and advance 
            PC. If the code object has no more instructions, return None.
        """
        if self.frame.pc >= len(self.frame.codeobject.code):
            return None
        else:
            instr = self.frame.codeobject.code[self.frame.pc]
            self.frame.pc += 1
            return instr
    
    def _is_in_toplevel_code(self):
        """ Is the VM currently executing the top-level code object?
        """
        return len(self.framestack) == 0
    
    def _create_global_env(self):
        global_binding = {}
        for name, func in builtins_map.items():
            global_binding[name] = BuiltinProcedure(name, func)

        # Add builtins which require access to the VM state.
        #
        global_binding['write'] = BuiltinProcedure('write', self._write)
        global_binding['debug-vm'] = BuiltinProcedure('debug-vm', self._hook_debug_vm)
        return Environment(global_binding)

    def _write(self, args):
        # Abides by the builtin procedure calling protocol - args is a Python
        # list of arguments.
        #
        self.output_stream.write(expr_repr(args[0]) + '\n')

    def _hook_debug_vm(self, args):
        print(self._show_vm_state())

    def _show_vm_state(self):
        def value_printer(item):
            if isinstance(item, Closure):
                return '| Closure <%s>\n' % item.codeobject.name
            elif isinstance(item, BuiltinProcedure):
                return '| BuiltinProcedure <%s>\n' % item.name
            else:
                return '| %s\n' % expr_repr(item)
        
        def frame_printer(item):
            str = 'Code: <%s>' % item.codeobject.name
            str += ' [PC=%s]\n' % item.pc
            return str
        
        def show_stack(stack, name, item_printer):
            str = ''
            head = '-' * (8 + len(name))
            str += '+%s+\n' % head
            str += '| %s stack |\n' % name
            str += '+%s+\n\n' % head
            
            i = 0
            while i < len(stack):
                str += '      |--------\n'
                if i == 0:
                    str += 'TOS:  '
                else:
                    str += '      '
                item = stack.peek(-1 - i)
                str += item_printer(item)
                i += 1
            str += '      |--------\n'
            return str

        str = show_stack(self.valuestack, 'Value', value_printer)
        str += '\n' + show_stack(self.framestack, 'Frame', frame_printer)
        return str


#------------------------------------------------------------------------------
if __name__ == '__main__':
    pass

File: experimental/barevm/src/vm.cpp
//*****************************************************************************
// bob: The virtual machine implementation
//
// Eli Bendersky (eliben@gmail.com)
// This code is in the public domain
//*****************************************************************************
#include "vm.h"
#include "utils.h"
#include "bytecode.h"
#include "environment.h"
#include "builtins.h"
#include "basicobjects.h"
#include <stack>
#include <algorithm>
#include <cstdio>
#include <stack>
#include <cassert>
#include <iostream>
#include <typeinfo>

using namespace std;


// Encapsulates the VM state - "execution frame". The frame consists of the
// current code object being executed, the pc (program counter) offset into
// it to know which instruction is next to execute, and the current
// environment in which the code object is being executed.
//
struct ExecutionFrame
{
    BobCodeObject* codeobject;
    unsigned pc;
    BobEnvironment* env;

    string repr()
    {
        return format_string("Code: <%s> [PC=%d]", codeobject->name.c_str(), pc);
    }
};


// A closure is a code object (procedure) with an associated environment
// in which the closure was created.
//
class BobClosure : public BobObject
{
public:
    BobClosure(BobCodeObject* codeobject_, BobEnvironment* env_)
        : codeobject(codeobject_), env(env_)
    {}

    virtual ~BobClosure()
    {}

    virtual string repr() const
    {
        return format_string("<closure '%s'>", codeobject->name.c_str());
    }

    BobCodeObject* codeobject;
    BobEnvironment* env;
};


struct VMImpl
{
    // The output stream for (write)
    //
    FILE* m_output_stream;

    // Explicit stack for execution frames to implement procedure calls.
    //
    stack<ExecutionFrame> m_framestack;

    // Stack for everything else
    //
    stack<BobObject*> m_valuestack;

    // The current execution frame
    //
    ExecutionFrame m_frame;

    //---------------------------------------------------------------

    // Builtins with access to VM state
    //
    BobObject* builtin_write(BuiltinArgs& args);
    BobObject* builtin_debug_vm(BuiltinArgs& args);

    // Internal 
    //
    BobEnvironment* create_global_env();
    string repr_vm_state();
};


BobVM::BobVM(const string& output_file)
    : d(new VMImpl)
{
    if (output_file == "") 
        d->m_output_stream = stdout;
    else {
        d->m_output_stream = fopen(output_file.c_str(), "w");

        if (!d->m_output_stream)
            throw VMError("Unable to open for output: " + output_file);
    }

    d->m_frame.codeobject = 0;
    d->m_frame.pc = 0;
    d->m_frame.env = d->create_global_env();
}


BobVM::~BobVM()
{
    if (d->m_output_stream != stdout)
        fclose(d->m_output_stream);
}


void BobVM::run(BobCodeObject* codeobj)
{
    if (!codeobj)
        return;

    d->m_frame.codeobject = codeobj;
    d->m_frame.pc = 0;

    // The big VM loop!
    //
    while (true) {
        BobCodeObject* cur_codeobj = d->m_frame.codeobject;

        // Get the next instruction from the current code object. If there's
        // no more instructions, this must be a top-level codeobject, in
        // which case the program is done.
        //
        BobInstruction instr;
        if (d->m_frame.pc >= cur_codeobj->code.size()) {
            if (d->m_framestack.size() == 0) 
                return;
            else
                throw VMError("Code object ended prematurely");
        }
        else {
            instr = cur_codeobj->code[d->m_frame.pc];
            d->m_frame.pc++;
        }

        switch (instr.opcode) {
            case OP_CONST:
            {
                assert(instr.arg < cur_codeobj->constants.size() && "Constants offset in bounds");
                BobObject* val = cur_codeobj->constants[instr.arg];
                cerr << "OP_CONST The repr of val on stack is " << val->repr() << endl;
                d->m_valuestack.push(val);
                break;
            }
            case OP_LOADVAR:
            {
                assert(instr.arg < cur_codeobj->varnames.size() && "Varnames offset in bounds");
                string varname = cur_codeobj->varnames[instr.arg];
                BobObject* val = d->m_frame.env->lookup_var(varname);
                if (!val)
                    throw VMError(format_string("Unknown variable '%s' referenced", varname.c_str()));
                d->m_valuestack.push(val);
                break;
            }
            case OP_STOREVAR:
            {
                assert(instr.arg < cur_codeobj->varnames.size() && "Varnames offset in bounds");
                assert(!d->m_valuestack.empty() && "Pop value from non-empty valuestack");
                BobObject* val = d->m_valuestack.top();
                d->m_valuestack.pop();
                string varname = cur_codeobj->varnames[instr.arg];
                BobObject* retval = d->m_frame.env->set_var_value(varname, val);
                if (!retval)
                    throw VMError(format_string("Unknown variable '%s' referenced", varname.c_str()));
                break;
            }
            case OP_DEFVAR:
            {
                assert(instr.arg < cur_codeobj->varnames.size() && "Varnames offset in bounds");
                assert(!d->m_valuestack.empty() && "Pop value from non-empty valuestack");
                BobObject* val = d->m_valuestack.top();
                d->m_valuestack.pop();
                d->m_frame.env->define_var(cur_codeobj->varnames[instr.arg], val);
                break;
            }
            case OP_POP:
            {
                // It's not a bug to generate instructions to pop the stack
                // when there's nothing to pop.
                //
                if (!d->m_valuestack.empty())
                    d->m_valuestack.pop();
                break;
            }
            case OP_JUMP:
            {
                d->m_frame.pc = instr.arg;
                break;
            }
            case OP_FJUMP:
            {
                assert(!d->m_valuestack.empty() && "Pop value from non-empty valuestack");
                BobBoolean* bool_predicate = dynamic_cast<BobBoolean*>(d->m_valuestack.top());
                d->m_valuestack.pop();
                if (bool_predicate && !bool_predicate->value())
                    d->m_frame.pc = instr.arg;
                break;
            }
            case OP_FUNCTION:
            {
                assert(instr.arg < cur_codeobj->constants.size() && "Constants offset in bounds");
                BobObject* val = cur_codeobj->constants[instr.arg];
                BobCodeObject* func_codeobj = dynamic_cast<BobCodeObject*>(val);
                assert(val && "Expected code object as the argument to OP_FUNCTION");
                d->m_valuestack.push(new BobClosure(func_codeobj, d->m_frame.env));
                break;
            }
            case OP_RETURN:
            {
                assert(!d->m_framestack.empty() && "OP_RETURN needs non-empty frame stack");
                d->m_frame = d->m_framestack.top();
                d->m_framestack.pop();
                break;
            }
            case OP_CALL:
            {
                // For OP_CALL we have the function on top of the value stack,
                // followed by its arguments (in reverse order). The amount of
                // arguments is in the argument of the instruction.
                // The function is either a builtin procedure or a closure.
                //
                assert(!d->m_valuestack.empty() && "Pop value from non-empty valuestack");
                BobObject* func_val = d->m_valuestack.top();
                d->m_valuestack.pop();
                vector<BobObject*> argvalues;

                // Take the function's arguments from the stack. The last
                // (right-most) argument is on top of the stack (first).
                //
                for (unsigned i = 0; i < instr.arg; ++i) {
                    assert(!d->m_valuestack.empty() && "Pop value from non-empty valuestack");
                    argvalues.push_back(d->m_valuestack.top());
                    d->m_valuestack.pop();
                }
                reverse(argvalues.begin(), argvalues.end());

                if (BobBuiltinProcedure* proc = dynamic_cast<BobBuiltinProcedure*>(func_val)) {
                    try {
                        BobObject* retval = proc->exec(argvalues);
                        d->m_valuestack.push(retval);
                    }
                    catch (const BuiltinError& err) {
                        throw VMError(err.what());
                    }
                }
                else if (BobClosure* closure = dynamic_cast<BobClosure*>(func_val)) {
                    // Extend the closure's environment with one where its code 
                    // object's arguments are bound to the values passed to it
                    // in the call.
                    //
                    if (argvalues.size() != closure->codeobject->args.size())
                        throw VMError(format_string("Calling procedure %s with %d args, expected %d",
                                        closure->codeobject->name.c_str(),
                                        argvalues.size(),
                                        closure->codeobject->args.size()));

                    BobEnvironment* call_env = new BobEnvironment(closure->env);
                    for (size_t i = 0; i < argvalues.size(); ++i) {
                        string argname = closure->codeobject->args[i];
                        BobObject* argvalue = argvalues[i];
                        call_env->define_var(argname, argvalue);
                    }

                    // To execute the procedure:
                    // 1. Save the current execution frame on the frame stack
                    // 2. Create a new frame from the closure's code object
                    //    and the extendend environment.
                    // 3. Start executing the frame by making it the current
                    //    frame with pc=0. The procedure's first instruction
                    //    will then execute in the next iteration of this 
                    //    loop.
                    //
                    d->m_framestack.push(d->m_frame);
                    ExecutionFrame new_frame;
                    new_frame.codeobject = closure->codeobject;
                    new_frame.pc = 0;
                    new_frame.env = call_env;
                    d->m_frame = new_frame;
                }
                else 
                    assert(0 && "Expected callable object on TOS for OP_CALL");

                break;
            }
            default:
                throw VMError(format_string("Invalid instruction opcode 0x%02X", instr.opcode));
        }
    }
}


// What follows is an attempt to create builtins that have access to the 
// internal state of the VM in a relatively clean way.
// The problem is that BuiltinProc is a pointer to a simple function that takes
// arguments and returns a value - it has no access to the VM. However, we
// also want to be able to create builtins that do have access to the VM.
// The way to do this is derive a special class from BobBuiltinProcedure.
// Instead of taking a BuiltinProc, BobVMBuiltinProcedure takes a VMImpl
// object and a pointer to a BobVM member function. 
// This way, by overriding exec(), it can be called in a way exactly
// similar to a normal BobBuiltinProcedure, but use its VM object reference
// and member function pointer to call a builtin function that's actually
// defined inside BobVM and thus has access to its state.
//
typedef BobObject* (VMImpl::*VMBuiltinProc)(BuiltinArgs& args);

class BobVMBuiltinProcedure : public BobBuiltinProcedure
{
public:
    BobVMBuiltinProcedure(const string& name, VMImpl& vmobj, VMBuiltinProc builtin)
        : BobBuiltinProcedure(name, 0), m_vmobj(vmobj), m_builtin(builtin)
    {}

    virtual ~BobVMBuiltinProcedure()
    {}

    virtual BobObject* exec(BuiltinArgs& args) const
    {
        // Invoke the member function via a pointer on the object
        //
        return (m_vmobj.*m_builtin)(args);
    }

private:
    VMImpl& m_vmobj;
    VMBuiltinProc m_builtin;
};


BobEnvironment* VMImpl::create_global_env()
{
    BuiltinsMap builtins_map = make_builtins_map();
    BobEnvironment* env = new BobEnvironment;

    // Add all the standard builtin procedures from the builtins module to the
    // environment
    //
    for (BuiltinsMap::const_iterator i = builtins_map.begin(); i != builtins_map.end(); ++i) {
        BobObject* proc = new BobBuiltinProcedure(i->first, i->second);
        env->define_var(i->first, proc);
    }

    // Now add the builtins defined as member functions of BobVM and have 
    // access to its state.
    //
    env->define_var("write", 
            new BobVMBuiltinProcedure("write", *this, &VMImpl::builtin_write));
    env->define_var("debug-vm", 
            new BobVMBuiltinProcedure("debug-vm", *this, &VMImpl::builtin_debug_vm));

    return env;
}


BobObject* VMImpl::builtin_write(BuiltinArgs& args)
{
    string output_str;

    cerr << "write called with n args: " << args.size() << endl;

    for (BuiltinArgsIteratorConst i = args.begin(); i != args.end(); ++i) {
        output_str += (*i)->repr() + " ";
    }
    output_str += "\n";
    fputs(output_str.c_str(), m_output_stream);

    return 0;
}


BobObject* VMImpl::builtin_debug_vm(BuiltinArgs& args)
{
    (void)args;
    string str = repr_vm_state();
    fputs(str.c_str(), m_output_stream);
    return 0;
}


static string frame_printer(ExecutionFrame frame)
{
    return "| " + frame.repr();
}


static string value_printer(BobObject* value)
{
    return "| " + value->repr();
}


template <class T>
static string repr_stack(stack<T> thestack, string name, string (*printer)(T))
{
    string head = string(8 + name.size(), '-');
    string str = format_string("+%s+\n| %s stack |\n+%s+\n\n",
                    head.c_str(), name.c_str(), head.c_str());

    // The stack is passed by value here (copied), so we're free to modify it.
    //
    bool tos = true;
    while (!thestack.empty()) {
        str += "     |--------\n";
        str += tos ? "TOS: " : "     ";
        str += printer(thestack.top()) + "\n";

        thestack.pop();
        tos = false;
    }
    str += "     |--------\n";

    return str;
}


string VMImpl::repr_vm_state()
{
    string str = repr_stack(m_valuestack, "Value", value_printer);
    str += "\n" + repr_stack(m_framestack, "Frame", frame_printer);
    return str;
}

